http-push.c

```diff
@@ -370,7 +370,7 @@ static void start_put(struct transfer_request *request)
 
 	/* Set it up */
 	git_deflate_init(&stream, zlib_compression_level);
-	size = git_deflate_bound(&stream, len + hdrlen);
+	size = git_deflate_bound(&stream, len + (size_t) hdrlen);
```

Comment:
We can't assume the type of len because it (git_deflate_bound) may be a macro, or a function, and it is then subject to the compile flags. 

Also, one of my readings of the various 'standards' suggested that it is within the implementation defined definition to downcast oversize variables if one (of the computation) was the normal 'right' size (e.g. long).

The whole up/down cast cascade and dual default function templates thing made me pedantic/explicit!

Reply:
> We can't assume the type of len because it (git_deflate_bound) may be a macro, or a function, and it is then subject to the compile flags.

Yes we can. Because we declared it ourselves, a couple lines above the shown diff context, as `size_t len`. So we can very much assume that its type is `size_t`.

> it is within the implementation defined definition to downcast oversize variables if one (of the computation) was the normal 'right' size (e.g. long).

Nope. Arithmetic expressions are always evaluated after upcasting narrower data types to larger ones. A compiler that downcasts is buggy.

You probably think about an example like this:

```c
short result;
long a = 1;
int b = 2;

result = a + b;
```

This is a bit tricky to understand, as the evaluation of the arithmetic expression `a + b` really upcasts the `int` to a `long` (unless they already have the same bit size). Only after that is the result downcast in order to be assigned to the narrower `result`.

That example has little relationship to the code under discussion, though, except in the case where `git_deflate_bound()` is defined as a function whose second parameter is declared with a too-narrow datatype. In which case you simply cannot do anything about it, except replace the function by our custom macro (as you did in another patch in this PR).

Reply:
I agree about the upcast to long. But when long is meant to be the maximum arithmetic calculation, then size_t may be down cast (as per Microsoft's desire that 32 bit code stays unchanged on 64 bit machines, so only (arithmetically) address 32bit range, unless one specially changes to full pointer calculation. As I understand it it's a bit open ended (which standard is your go to reference document?)

It maybe that the commit message needs to be long to explain the ripple through consequences, or we make the code informative. Its one of those 'discursive' areas.

Reply:
> We can't assume the type of len

But we can. It is _declared_ as `size_t len;`. See here: https://github.com/git-for-windows/git/blob/a4b2bb588bb9680e4ac2ce979cbf954f582cb9df/http-push.c#L363